File: libreastream/src/main/java/com/eje_c/libreastream/ReaStreamReceiver.java
package com.eje_c.libreastream;

import java.io.IOException;
import java.net.DatagramPacket;
import java.net.DatagramSocket;
import java.net.SocketException;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;

public class ReaStreamReceiver implements AutoCloseable {

    public static final int DEFAULT_PORT = 58710;
    public static final String DEFAULT_IDENTIFIER = "default";
    private final DatagramSocket socket;
    private final DatagramPacket packet;
    private final ReaStreamPacket audioPacket = new ReaStreamPacket();
    private final ByteBuffer buffer;
    private String identifier = DEFAULT_IDENTIFIER;

    /**
     * Create ReaStream receiver with UDP socket default port.
     *
     * @throws SocketException
     */
    public ReaStreamReceiver() throws SocketException {
        this(DEFAULT_PORT);
    }

    /**
     * Create ReaStream receiver with UDP socket specific port.
     *
     * @param port UDP port
     * @throws SocketException
     */
    public ReaStreamReceiver(int port) throws SocketException {
        this(new DatagramSocket(port));
    }

    /**
     * Create ReaStream receiver with UDP socket specific port.
     *
     * @param socket Pre-created socket
     */
    public ReaStreamReceiver(DatagramSocket socket) {
        buffer = ByteBuffer.allocate(ReaStreamPacket.MAX_BLOCK_LENGTH + ReaStreamPacket.AUDIO_PACKET_HEADER_BYTE_SIZE);
        buffer.order(ByteOrder.LITTLE_ENDIAN);
        packet = new DatagramPacket(buffer.array(), buffer.capacity());
        this.socket = socket;
    }

    /**
     * @param identifier Identifier
     */
    public void setIdentifier(String identifier) {
        this.identifier = identifier;
    }

    /**
     * @return Identifier
     */
    public String getIdentifier() {
        return identifier;
    }

    /**
     * Wait for receiving {@link ReaStreamPacket}.
     * This method blocks until receiving ReaStream packet with same identifier.
     *
     * @return Received ReaStream packet.
     * @throws IOException
     */
    public ReaStreamPacket receive() throws IOException {

        do {
            socket.receive(packet);
            audioPacket.readFromBuffer(buffer);
        } while (!identifier.equals(audioPacket.getIdentifier()));

        return audioPacket;
    }

    /**
     * Sets the read timeout in milliseconds for this socket.
     * This receive timeout defines the period the socket will block waiting to
     * receive data before throwing an {@code InterruptedIOException}. The value
     * {@code 0} (default) is used to set an infinite timeout. To have effect
     * this option must be set before the blocking method was called.
     *
     * @param timeout the timeout in milliseconds or 0 for no timeout.
     * @throws SocketException if an error occurs while setting the option.
     */
    public void setTimeout(int timeout) throws SocketException {
        socket.setSoTimeout(timeout);
    }

    /**
     * Close UDP socket.
     */
    @Override
    public void close() {
        socket.close();
    }
}

File: libreastream/src/main/java/com/eje_c/libreastream/ReaStreamReceiver.kt
package com.eje_c.libreastream

import java.net.InetSocketAddress
import java.net.SocketException
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.nio.channels.DatagramChannel
import java.nio.channels.SelectionKey
import java.nio.channels.Selector
import kotlin.concurrent.thread

/**
 * Low-level [ReaStreamPacket] receiver.
 *
 * @param channel Pre-created [DatagramChannel] for reuse channel. If [close] is called, this channel is also closed.
 * @param identifier Identifier string. Max 32 bytes. Default value is "default".
 * @param port Waiting port number. Default value is [ReaStream.DEFAULT_PORT] which is used by original ReaStream plugin.
 */
class ReaStreamReceiver(
        private val channel: DatagramChannel = DatagramChannel.open(),
        val identifier: String = ReaStream.DEFAULT_IDENTIFIER,
        port: Int = ReaStream.DEFAULT_PORT) : AutoCloseable {

    private val reaStreamPacket = ReaStreamPacket()
    private val buffer: ByteBuffer = ByteBuffer.allocate(ReaStreamPacket.MAX_BLOCK_LENGTH + ReaStreamPacket.AUDIO_PACKET_HEADER_BYTE_SIZE)
            .apply {
                order(ByteOrder.LITTLE_ENDIAN)
            }
    private var closed: Boolean = false

    var onReaStreamPacketListener: OnReaStreamPacketListener? = null

    init {
        channel.socket().bind(InetSocketAddress(port))
        channel.configureBlocking(false)

        // Start async receiving thread
        thread {
            val selector = Selector.open()
            channel.register(selector, SelectionKey.OP_READ)

            // Stop when close() is called
            while (!closed) {

                selector.select()
                val iterator = selector.selectedKeys().iterator()

                while (iterator.hasNext()) {
                    val key = iterator.next()
                    iterator.remove()

                    if (!key.isValid) {
                        continue
                    }

                    if (key.isReadable) {
                        buffer.clear()
                        channel.receive(buffer)
                        reaStreamPacket.readFromBuffer(buffer)
                        if (identifier == reaStreamPacket.getIdentifier()) {
                            onReaStreamPacketListener?.onReceive(reaStreamPacket)
                        }
                    }
                }
            }
        }

    }

    /**
     * Sets the read timeout in milliseconds for this channel.
     * This receive timeout defines the period the channel will block waiting to
     * receive data before throwing an `InterruptedIOException`. The value
     * `0` (default) is used to set an infinite timeout. To have effect
     * this option must be set before the blocking method was called.
     *
     * @param timeout the timeout in milliseconds or 0 for no timeout.
     * @throws SocketException if an error occurs while setting the option.
     */
    @Throws(SocketException::class)
    fun setTimeout(timeout: Int) {
        channel.socket().soTimeout = timeout
    }

    /**
     * Close UDP channel.
     */
    override fun close() {
        closed = true
        channel.close()
    }
}
